signer/signer.go

```diff
@@ -301,6 +325,81 @@ func parseDSAPKCS8PrivateKey(der []byte) (*dsa.PrivateKey, error) {
 	}, nil
 }
 
+// MakeKey generates a new key of type keyTpl and returns the priv and public interfaces.
+// If an HSM is available, it is used to generate and store the key, in which case 'priv'
+// just points to the HSM handler and must be used via the crypto.Signer interface.
+func (cfg *Configuration) MakeKey(keyTpl interface{}, keyName string) (priv crypto.PrivateKey, pub crypto.PublicKey, err error) {
+	if cfg.isHsmAvailable {
+		var slots []uint
+		slots, err = cfg.hsmCtx.GetSlotList(true)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to list PKCS#11 Slots")
+		}
+		if len(slots) < 1 {
+			return nil, nil, errors.New("failed to find a usable slot in hsm context")
+		}
+		switch keyTpl.(type) {
+		case *ecdsa.PublicKey:
+			priv, err = crypto11.GenerateECDSAKeyPairOnSlot(slots[0], []byte(keyName), []byte(keyName), keyTpl.(*ecdsa.PublicKey))
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to generate ecdsa key in hsm")
+			}
+			pub = priv.(*crypto11.PKCS11PrivateKeyECDSA).Public()
+			return
+		case *rsa.PublicKey:
+			keySize := keyTpl.(*rsa.PublicKey).Size()
+			priv, err = crypto11.GenerateRSAKeyPairOnSlot(slots[0], []byte(keyName), []byte(keyName), keySize)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to generate rsa key in hsm")
+			}
+			pub = priv.(*crypto11.PKCS11PrivateKeyRSA).Public()
+			return
+		default:
+			return nil, nil, errors.Errorf("making key of type %T is not supported", keyTpl)
+		}
+	}
+	// no hsm, make keys in memory
+	switch keyTpl.(type) {
+	case *ecdsa.PublicKey:
+		switch keyTpl.(*ecdsa.PublicKey).Params().Name {
+		case "P-256":
+			priv, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		case "P-384":
+			priv, err = ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+		default:
+			return nil, nil, fmt.Errorf("unsupported curve %q",
+				keyTpl.(*ecdsa.PublicKey).Params().Name)
+		}
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to generate ecdsa key in memory")
+		}
+		pub = priv.(*ecdsa.PrivateKey).Public()
+		return
+	case *rsa.PublicKey:
+		keySize := keyTpl.(*rsa.PublicKey).Size()
+		priv, err = rsa.GenerateKey(rand.Reader, keySize)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to generate rsa key in memory")
+		}
+		pub = priv.(*rsa.PrivateKey).Public()
+		return
+	default:
+		return nil, nil, errors.Errorf("making key of type %T is not supported", keyTpl)
+	}
+}
+
+// GetPrivKeyHandle returns the hsm handler object id of a key stored in the hsm,
+// or 0 if the key is not stored in the hsm
+func GetPrivKeyHandle(priv crypto.PrivateKey) uint {
+	switch priv.(type) {
+	case *crypto11.PKCS11PrivateKeyECDSA:
+		return uint(priv.(*crypto11.PKCS11PrivateKeyECDSA).Handle)
```

Comment:
nit: I think we can do something to avoid the second type assertion like:

```golang
key := switch priv.(type) {
case *crypto11.PKCS11PrivateKeyECDSA:
		return uint(key.Handle)
...
}
```

Reply:
that syntax doesn't seem to work:
```go
func GetPrivKeyHandle(priv crypto.PrivateKey) uint {
	key := switch priv.(type) {
	case *crypto11.PKCS11PrivateKeyECDSA, *crypto11.PKCS11PrivateKeyRSA:
		return uint(key.Handle)
	}
	return 0
}
```
```
# go.mozilla.org/autograph/signer
signer/signer.go:397:9: syntax error: unexpected switch, expecting expression
signer/signer.go:399:3: syntax error: unexpected return at end of statement
signer/signer.go:401:2: syntax error: non-declaration statement outside function body
```

Reply:
oh I think we need to handle each case individually like the second example: https://github.com/golang/go/wiki/Switch#type-switch